### ISSUE TITLE
Enhancement: New Inline Asm Syntax

### ISSUE THREAD

**Issue by EleanorNB** [ziglang/zig#5241]
# New Inline Asm Syntax
`asm (arches) (bindings, clobbers) (:return_register|void|noreturn) { local_labels body } (else ...)?` + `config?` (somewhere)

## Arches
An optional list of target architectures. If this is `null`, the block is assumed to be for all architectures (an assembler error is always a compile error). Otherwise, one of these must match `builtin.arch`, or an `else` branch must be present. This is a list rather than a single value as some architectures have mutually compatible subsets (e.g. 8086/x86/x86_64, MIPS/RISC-V).

## Bindings and Clobbers
Bindings have the form `"register" name: type = init_value`. `name` can be `_`, if the register is desired only for initialisation. `name` can also be a variable in scope, in which case `type` and `init_value` are omitted, and changes to this register's value are taken as changes to the variable. `init_value` can be `undefined`, in which case `type` can be omitted (it doesn't matter much in assembly anyway), unless `name` is the return register (more on this later). Clobbers are simply `"register"`.

## Return Register
A binding can be nominated as the return value, with `:name`. (Allowing `:"register"` would cause parsing ambiguity, and this can be trivially done with a binding anyway.) `void` and `noreturn` are also allowed. Reaching the end of a `noreturn` block is safety-checked UB.

## Local Labels
A list of local labels. Formatted as strings.

Local labels are unique to the block: `%(label)` matches `%(label)` within the block, and is guaranteed not to match anything else in the program. They are listed within the braces of the body because they really don't make sense outside that context.

## Body
The assembly code itself, as a string. If this fails to assemble, it's a compile error.

The following macros are defined:
- `%[name]`
Register, as specified in bindings section.
- `%(label)`
Label, as listed in local labels section.
- `@[variable]`
Pre-mangled global variable name. Used to reference globals. See #5211.
- `@(function)`
Pre-mangled function name. Used to call functions. See #5211.

A literal `%` or `@` is escaped with another one: `%%` or `@@`. Strictly speaking, if we're substituting text, only one of `@[]` and `@()` is needed -- but, if we want to integrate the assembler with the compiler, the distinction may be important, so I've listed both.

## Else
If `arches` is non-null and none of the listed architectures match `builtin.arch`, this is compiled instead. Can be used to switch on architectures, optimise a specific architecture only, or simply `@compileError`. If this is not present, a target mismatch is a compile error.

**N.B.**: An `else` branch is only allowed if `arches` is non-null. This decision was made because, when you set `arches` to `null`, either you know execution will never reach this point on the wrong architecture, or you only care about compiling for a specific architecture. In the former case, you definitely want an unexpected architecture to be a compile error; and in the latter, to support a new architecture, the laziest thing you can do is start caring.

## Config
Configuration is passed in a pragma (#5239) with the following fields:
- `impure`
This block has side effects.
- `stack(n)`
This block allocates `n` bytes on the stack. Defaults to `0`.
- `calls(funcs)`
This block calls the functions listed in `funcs`. Defaults to `.{}`.

## Example
``` zig
const builtin = @import("builtin");

const fib_asm = fn (n: u32) u32 {
    return asm (.{ builtin.Arch.riscv64, builtin.Arch.riscv32 }) @{
        stack(12),
        calls(.{ fib_iter }),
    } (
        "a0" this  : u32 = 0,
        "a1" next  : u32 = 1,
        "=r" to_go : u32 = n,
    ) :this {
        .{ "loop", "end" }

        \\%(loop):
        \\  bez %[to_go], %(end)
        // We can do function pro/epi at callsite!
        \\  addi sp, -12
        \\  sd ra, 0(sp)
        \\  sw %[to_go], 8(sp)
        \\  call @(fib_iter)
        \\  lw %[to_go], 8(sp)
        \\  ld ra, 0(sp)
        \\  addi sp, 12
        \\  addi %[to_go], -1
        \\  j %(loop)
        \\%(end):
    } else @compileError("Your machine could be better");
};

// Actually returns two values, but the compiler has no way to express that
const fib_iter = fn @{callconv(.Naked)} (this: u32, next: u32) void {
    // No need to check architecture -- we'll only call this from fib_asm
    asm (null) @{impure} (
        "a0" this,
        "a1" next,
        "=r" temp = undefined,
    ) void {
        .{}

        \\  add %[temp], %[this], %[next]
        \\  mv %[this], %[next]
        \\  mv %[next], %[temp]
    };
};
```

## TL;DR: Benefits over Status Quo
- If any of the sections are missed, the compiler can detect exactly which ones
- Order of mandatory components has a logical progression, just like function declaration
- Option to tie to target architecture
- Registers have types
- Can express non-returning and valueless assembly
- Can reference global variables and call functions
- Won't unexpectedly jump to random points in the program
- Communicates metadata to compiler, but does not require it
- Provides alternative for unsupported architectures
- Can be automatically distinguished from status quo, albeit with some lookahead
- Can be automatically derived from status quo

_Originally posted by @EleanorNB in https://github.com/ziglang/zig/issues/215#issuecomment-622244402_

**Comment by EleanorNB** [comment]
Hey @andrewrk, we want to stabilise the language, right? So a decision on this may be pertinent.

**Comment by lerno** [comment]
Compared to GCC style syntax this is much more verbose. So would it really make sense to do this?

What people *really* like is MSVC style, where clobbers, register allocation etc are mostly inferred by the compiler. This requires lots of good defaults, but is great to work with: downside is a lot of loss of control, which could be regained with some clever added optional constraints instead. That said it might violate the Zig explicitness goal.

So if the ideal isn't achievable, why pick a style that is completely new for Zig, rather than the de facto standard?

**Comment by lerno** [comment]
Also, everything is still strings, so it's basically leaving everything stringly typed. Compare to the Rust asm that at least defines what looks as constants for things like `xmm_reg` behaves as an actual constant rather than a random string. Similarly registers in this proposals are also just strings.

I suspect people will just consider this a confusing, hobbled and verbose version of GCC inline asm.

Compare this:

```zig
return asm {
        "eax": u32 = 0,
        "ebx" b: u32,
        "ecx" c: u32,
        "edx" d: u32,
        "?memory",
    } "cpuid"
    : .{ b, c, d };
```
To
```c
static inline void cpuid(int code, uint32_t* a, uint32_t* d)
{
    asm volatile ( "cpuid" : "=a"(*a), "=d"(*d) : "0"(code) : "ebx", "ecx" );
}
```
You don't like that? Here's another one using code = 1 without the memory:
```c
int a = 0x1, b, c, d;
asm ( "cpuid"  : "=a" (a), "=b" (b), "=c" (c), "=d" (d) : "0" (a) );
```
So what is the benefit?

**Comment by lerno** [comment]
For comparison, the Rust new asm: https://doc.rust-lang.org/beta/unstable-book/library-features/asm.html

**Comment by N00byEdge** [comment]
I like this syntax. A lot. But there is one issue here I think just looks a little weird to me

> ```
> const _start = fn callconv(.naked) () noreturn {
>     if (builtin.os.tag == .wasi) {
>         std.os.wasi.proc_exit(@call(.{ .modifier = .always_inline }, callMain, .{}));
>     }
> 
>     asm {
>         "?reg" stack_ptr: [*]usize,
>     // Much more compact and local
>     } switch (builtin.arch) {
>         .x86_64 => "mov ?(stack_ptr), rsp",
>         .i386 => "mov ?(stack_ptr), esp",
>         .aarch64, .aarch64_be, .arm => "mov ?(stack_ptr), sp",
>         .riscv64 => "mv ?(stack_ptr), sp"
>         .mips, .mipsel => (
>           \\ .set noat
>           \\ move ?(stack_ptr), $sp
>         ),
>         else => @compileError("unsupported arch"),
>     }
>     // By the time we get here, we have the stack pointer
>     // -- so, no global required
>     : @call(.{ .modifier = .never_inline }, posixCallMainAndExit, .{ stack_ptr });
> };
> ```

In here, I just want to grab the value of a register. I don't care about how the mov instruction, and I believe that should be left to the compiler to figure out. Should putting empty asm and replacing `"?reg"` with a switch on the arch returning `"rsp"` etc be allowed instead?

**Comment by EleanorNB** [comment]
Hmm, I hadn’t thought of that. My instinct is to allow this, but I’m not sure if this would lead to parsing ambiguity. If not, then sure.

**Comment by ethindp** [comment]
I have an alternative proposal that, I think, will be much clearer, and far different from GCC inline asm, or any other asm syntax I've seen.  And it won't be just string hackery, either. I imagine this proposal will take a long time to actually implement, but it'll be much, much clearer, and very elegant, and fits Zig's Zen (whereas the current proposal doesn't).

The general idea are `asm` blocks. The syntax is similar, but with some significant differences:

* As is currently done, `asm` blocks begin with the keyword `asm`, followed by the optional keyword `volatile`. Then, either:
    - a parenthesized list of semicolon-separated inputs, outputs, and clobber specifications followed by a block; or
    - a block.
    * Inputs, outputs, and clobbers are specified before any assembly statements. (I use "assembly statements" here deliberately, see below.) Input specifications take the form `inputs: element1, element2, element3, ...`; output and clobber specifications are similar but using the keyword `outputs` or `clobbers`, respectively.
    * Input, output, and clobber specifications take input, output, or clobber elements. An input, output, or clobber element can be either of the following:
        - the form `arg = value`, where `arg` can be a register or variable, and value can either be a register, variable, or 'memory'; or
        - the form `value`, which can either be a register or variable, or for the case of clobbers, 'memory'.
    * In the second specification form, the (actual) value is the only way to refer to said value in assembly statements; the first form could be considered a "renaming" of the item.
    * The input/output/clobber specifications are optional, but if parentheses come before the block, at least one of those specifications must be provided. There can only be one of all of the specifications in any given `asm` block.
    * If the input or output is a compound data type (array, slice, struct, union, ...), that entire compound data type is considered as the input or output; you cannot solely use as an input or output a constituant field or element of that data type.
* The block contains assembly statements. An assembly statement can either be an assignment statement or instruction.
* In the case of an assignment statement, the form is `a = b;`, just as in zig. However, assignments must be "split" assignments; that is, you cannot do `a[2..5] = 3;`. This is because assignment statements are directly translated into loads and stores, and this version of the syntax doesn't allow for multi-loads and stores because that would be quite complex depending on the architecture, and this proposal is already complex as is. Perhaps we can change this in the future, but I think this is an acceptable limitation for now.
* The LHS of an assignment statement must be a register, dereference, or constituant element or field of a compound data type. You cannot write to arbitrary memory using this construction; for that, you have to use actual instructions. This is mainly because allowing arbitrary address writes and reads would look quite odd (at least in my opinion), and the usual way of doing this is to load the address into a register and then write to it that way.
* In the case of instructions, these look like function calls. This sticks to Zigs "favor reading code over writing code" ideal, and also makes things easier for people who aren't experts with inline assembly. For example, the instruction `vpcmpltud k3, ymm3, ymm0` would be translated into `vpcmpltud(k3, ymm3, ymm0);`. Similarly, the ARM instruction `LDR r0, [r1]` would be translated into `ldr(r0, &r1);`. (I'm unsure how to translate an instruction like `STMFD sp!, {r0-r3, lr}` into this syntax, and would appreciate assistance to refine it.)
* Labels have the same syntax as in zig; same for referring to them in assembly statements (e.g. `jmp :do_something`).

To provide an example in action, here's the classic CPUID on x86, from Agner Fog's asmlib library, which uses a parameter as a return value (but for this example we just allocate it on the stack and use that). The original example is as follows:

```asm
cpuid_ex:
%IFDEF   WINDOWS
; parameters: rcx = abcd, edx = a, r8d = c
        push    rbx
        xchg    rcx, r8
        mov     eax, edx
        cpuid                          ; input eax, ecx. output eax, ebx, ecx, edx
        mov     [r8],    eax
        mov     [r8+4],  ebx
        mov     [r8+8],  ecx
        mov     [r8+12], edx
        pop     rbx
%ENDIF        
%IFDEF   UNIX
; parameters: rdi = abcd, esi = a, edx = c
        push    rbx
        mov     eax, esi
        mov     ecx, edx
        cpuid                          ; input eax, ecx. output eax, ebx, ecx, edx
        mov     [rdi],    eax
        mov     [rdi+4],  ebx
        mov     [rdi+8],  ecx
        mov     [rdi+12], edx
        pop     rbx
%ENDIF        
        ret
```

We'll drop the prologue, and the example in this proposed syntax becomes:

```zig
fn cpuid(a: u32, c: u32) [4]u32 {
    var abcd: [4]u32 = undefined;
    asm(inputs: a, c; outputs: abcd; clobbers: eax, ecx) {
        edx = eax;
        cpuid();
        // These are memory-based movs
        abcd[0] = eax;
        abcd[1] = ebx;
        abcd[2] = ecx;
        abcd[3] = edx;
    }
    return abcd;
}
```

As another example, take a more complex one, loading the GDT (sorry if this isn't quite valid, I'm not the most skilled at this):

Original:

```asm
load_gdt:
    push %rbp
    mov %rsp, %rbp
    sub $32, %rsp
    mov 8(%rsp), %rax
    lgdt (%rax)
    pushq $0x08
    lea reload_segment_regs(%rip), %rax
    push %rax
    lretq
reload_segment_regs:
    mov $0x10, %ax
    mov %ax, %ds
    mov %ax, %es
    mov %ax, %fs
    mov %ax, %gs
    mov %ax, %ss
    mov %rbp, %rsp
    pop %rbp
    ret
```

In this syntax, this becomes:

```zig
fn load_gdt(gdt: usize) void {
    asm(inputs: gdt) {
        rax = gdt;
        lgdt(&rax);
        push(0x08);
        lea(:reload_segment_regs); // labels are always PIC/PIE unless `build.zig` explicitly indicates that the executable is not position independent
        push(rax);
        lret(); // long return
        reload_segment_regs:
        // Register-immediate load
            ax = 0x10;
            // register-register load and store
            ds = ax;
            es = ax;
            fs = ax;
            gs = ax;
            ss = ax;
    }
    }
```

Like I said, this definitely needs refinement and I think that this will take a long time to completely implement. However, I think that this is, most likely, the proposal that upholds Zig's zen and doesn't make inline assembly look like a complete and utter mess. This syntax has the benefit of giving the compiler a lot of information about what your trying to do, so it could very well optimize your loads/stores into something using AVX or neon if possible. What I'm unsure about are things like:

- explicit pointer size indicators (e.g. `dword ptr`)
- instruction prefixes (`rex64`, `rex.w`, etc.) (though we could perhaps make these just another function call)
- ARM ranged loads/stores
- AVX masking and broadcasting (e.g. `vmovdqu8 zmm16{k1}{z}, [rsi]` or `vpaddd  ymm4 {k2}, ymm4, dword ptr [ADD1] {1to8}`)
- Memory offset (`mov rbx, [rax + 0x32]` for instance) (though maybe we could do `mov(rbx, &(rax op...));`?)

If you guys want to help refine this I'd appreciate it. I know that some of the syntactic elements that this introduces are unorthodox, and are quite different from Zig's normal syntax, but I did try to stay as close to Zig as possible while compromising on the fact that this was inline assembly and I didn't really have much of a choice. For the RHS of an assignment statement, most valid expressions are allowed, barring multi-loads or stores; I was thinking that you could even call built-ins as well. When this happened, the load/store would be a multi-load/store, but would finish as a single load/store; for example, if you used `xmm0 = @sqrt(...))`, and `@sqrt` resulted in the `VSQRTSD` instruction, the compiler would translate your code appropriately to execute `VSQRTSD` and then would do a final `load` into `xmm0`. Conversely, if it resulted in a libc function call, the compiler would issue the appropriate instructions for a function call, then (attempt) to store the result in `xmm0`; if it couldn't, an error would result.

I understand that this syntax would result in "behind your back" instructions in certain instances. In the case of the aforementioned built-in function call idea, discarding that for now would be perfectly reasonable. The assignment statement thing was to eliminate the minutia of a ton of `mov`s, or whatever the target uses for loads and stores, and to instead allow the programmer to focus on what they really are using inline assembly to accomplish. Obviously, if you really wanted to they could fall back to setting everything up themselves; this syntax does not prevent you from using any instruction that the target supports, even if there is a more "abstract" syntax available.

**Comment by lerno** [comment]
@ethindp You might want to draw some inspiration from how C3 does it: https://c3-lang.org/asm/ It creates a very simple, regular grammar and infers clobbers.

**Comment by ethindp** [comment]
@lerno That's an interesting syntax, but IMO it's not as clear as mine (but mine is more complex since I'm trying to be as flexible as possible).

**Comment by lerno** [comment]
@ethindp Yes, the focus is trying to be as cheap as possible to implement for various variants of asm.
